### PR TITLE
Quota enforce at sector allocator

### PR DESCRIPTION
**Pull request by meroton-benjamin**

### CODE REVIEW

**Comment by meroton-benjamin** [comment]
Latest version of an implementation of a quota enforcing sector allocator. There are some possible performance improvements that have not been pursued:

**GetNextUnmappedSector** when run in a dense region iterates through all sectors until it finds an unmapped one, this is not worse than what was done before but if run on e.g. a 16TB dense file from byte zero would require 2^32 iterations until it found an unmapped sector. With some bookkeeping this could probably be reduced.

**GetLogicalSize** keeps track of the last mapped sector of the pointer by manipulating an internal variable. This needs to be recalculated whenever truncated to smaller than the mapped sector. With a better implementation of GetNextUnmappedSector this might be possible to do faster by alternating between GetNextUnmappedSector and GetNextMappedSector until it reaches the last mapped sector less than what was truncated to.

File: pkg/filesystem/pool/sector_pointer.go [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+// directSectors, singleIndirectionSectors, doubleIndirectionSectors,
+// and tripleIndirectionSectors are the maximum number of sectors which
+// can be represented by a maximum of 0, 1, 2, and 3 indirections
+// respectively.
+const (
+	indirectionArraySize     = 1626 // Smallest size where the full unsigned 32 bit range can be addressed.
+	directSectors            = 12
+	singleIndirectionSectors = directSectors + indirectionArraySize
+	doubleIndirectionSectors = singleIndirectionSectors + indirectionArraySize*indirectionArraySize
+	tripleIndirectionSectors = doubleIndirectionSectors + indirectionArraySize*indirectionArraySize*indirectionArraySize
```
Comment (EdSchouten) [inline pkg/filesystem/pool/sector_pointer.go:54]: Can you please put this underneath?
```go
_ uint = tripleIndirectionSectors - 1<<32
```
This will throw a compile-time error if `indirectionArraySize` is too small.

File: pkg/filesystem/pool/sector_pointer.go [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+//   - No indirection for files up to ~40KiB
+//   - Single indirection for files up to ~4MiB
+//   - Double indirection for files up to ~4GiB
+//   - Triple indirection for files up to ~4TiB
+type SectorPointer struct {
+	direct              [12]uint32
+	single              *indirect[uint32]
+	double              *indirect[*indirect[uint32]]
+	triple              *indirect[*indirect[*indirect[uint32]]]
+	logicalSectorLength uint32
```
Comment (EdSchouten) [inline pkg/filesystem/pool/sector_pointer.go:40]: I've got a gut feeling that we should be able to get rid of this field...

File: pkg/filesystem/pool/sector_pointer.go [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+type indirect[T any] struct {
+	val [indirectionArraySize]T
+}
+
+// GetNextMappedSector returns the logical index of the next mapped
+// sector in the sector map. If the start index is mapped then start is
+// returned. If there are no more mapped sectors then io.EOF is
+// returned.
+func (sp *SectorPointer) GetNextMappedSector(logical uint32) (uint32, error) {
+	from := logical
```
Comment (EdSchouten) [inline pkg/filesystem/pool/sector_pointer.go:66]: I think this code is a bit hard to read due to the excessive use of helper variables. What are your thoughts on getting rid of `next` and `from`? Can't we just write:
```go
for {
    var list []uint32
    logical, list = sp.getNextDirectSectorList(logical)
    ...
}
```

File: pkg/filesystem/pool/block_device_backed_file_pool.go [hunk 4]
```diff
@@ -35,22 +35,21 @@ func NewBlockDeviceBackedFilePool(blockDevice blockdevice.BlockDevice, sectorAll
...
 	}, nil
 }
 
 type blockDeviceBackedFile struct {
-	fp        *blockDeviceBackedFilePool
-	sizeBytes uint64
-	sectors   []uint32
+	fp           *blockDeviceBackedFilePool
+	sizeBytes    uint64
+	sectorMapper *SectorPointer
```
[the remark this answers is not in the document]
Comment (meroton-benjamin) [reply pkg/filesystem/pool/block_device_backed_file_pool.go]: SectorPointer renamed to SectorMap and the SectorMap struct (70 bytes) was added directly to the File objects

File: pkg/filesystem/pool/configuration.go [hunk 5]
```diff
@@ -12,32 +11,59 @@ import (
...
+// their own SectorAllocators. This is required to manage quotas for
+// different file pools.
+type FilePoolFactory interface {
+	NewFilePool(allocator SectorAllocator) FilePool
+}
+
+type filePoolFactory struct {
+	blockDevice     blockdevice.BlockDevice
+	sectorSizeBytes int
+}
```
[the remark this answers is not in the document]
Comment (meroton-benjamin) [reply pkg/filesystem/pool/configuration.go]: Ack

File: pkg/filesystem/pool/quota_enforcing_file_pool.go [hunk 6]
```diff
@@ -1,55 +1,28 @@
...
-func (m *quotaMetric) release(v int64) {
-	m.remaining.Add(v)
-}
-
 type quotaEnforcingFilePool struct {
 	base FilePool
 
-	filesRemaining quotaMetric
-	bytesRemaining quotaMetric
+	filesRemaining *quotaMetric
```
[the remark this answers is not in the document]
Comment (meroton-benjamin) [reply pkg/filesystem/pool/quota_enforcing_file_pool.go]: `quotaMetric` contains an `atomic.Int64` which shouldn't be copied after it has been used. To accomodate the quota metric as an inline struct I have moved initialization away from the `newQuotaMetric` into an `init` method.

File: pkg/filesystem/pool/sector_pointer.go [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+	"math"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// NewSectorPointer returns a new SectorPointer struct.
+func NewSectorPointer() *SectorPointer {
+	return &SectorPointer{}
+}
```
[the remark this answers is not in the document]
Comment (meroton-benjamin) [reply pkg/filesystem/pool/sector_pointer.go]: Ack

File: pkg/filesystem/pool/sector_pointer.go [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+// Otherwise the value corresponds to the sector index where the data
+// can be found.
+//
+// The larger the file the more indirections it allows. For a typical
+// 4096 byte sector size this results in:
+//   - No indirection for files up to ~40KiB
+//   - Single indirection for files up to ~4MiB
+//   - Double indirection for files up to ~4GiB
+//   - Triple indirection for files up to ~4TiB
+type SectorPointer struct {
```
[the remark this answers is not in the document]
Comment (meroton-benjamin) [reply pkg/filesystem/pool/sector_pointer.go]: Ack

File: pkg/filesystem/pool/sector_pointer.go [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+	direct              [12]uint32
```
[the remark this answers is not in the document]
Comment (meroton-benjamin) [reply pkg/filesystem/pool/sector_pointer.go]: ack

File: pkg/filesystem/pool/block_device_backed_file_pool.go [hunk 10]
```diff
@@ -62,44 +61,51 @@ func (f *blockDeviceBackedFile) toDeviceOffset(sector uint32, offsetWithinSector
...
-	firstSectorIndex := int(off / int64(f.fp.sectorSizeBytes))
-	endSectorIndex := int((uint64(off) + uint64(n) + uint64(f.fp.sectorSizeBytes) - 1) / uint64(f.fp.sectorSizeBytes))
-	if endSectorIndex > len(f.sectors) {
-		endSectorIndex = len(f.sectors)
+func (f *blockDeviceBackedFile) getInitialSectorIndex(off int64, n int) (uint32, uint32, int) {
+	firstSectorIndex := uint32(off / int64(f.fp.sectorSizeBytes))
+	endSectorIndex := uint32((uint64(off) + uint64(n) + uint64(f.fp.sectorSizeBytes) - 1) / uint64(f.fp.sectorSizeBytes))
+	if endSectorIndex > f.sectorMapper.GetLogicalSize() {
+		endSectorIndex = f.sectorMapper.GetLogicalSize()
 	}
```
[the remark this answers is not in the document]
Comment (meroton-benjamin) [reply pkg/filesystem/pool/block_device_backed_file_pool.go]: After looking deeper at `block_device_backed_file_pool.go` I could remove the uses of `GetLogicalSize` fairly easily. I have removed the underlying method as well.

File: pkg/filesystem/pool/sector_pointer.go (hunk 2 above)
Comment (meroton-benjamin) [reply pkg/filesystem/pool/sector_pointer.go]: done

File: pkg/filesystem/pool/sector_pointer.go (hunk 3 above)
Comment (meroton-benjamin) [reply pkg/filesystem/pool/sector_pointer.go]: I don't really find this suggestion making things any clearer, I appreciate that the why of `GetNextMappedSector`  is not very clear unless you know how the helper `getNextDirectSectorList` works. I have added some documentation to `getNextDirectSectorList` and reworked it somewhat to get around an overflow bug. Please have a look again if you still find it too unclear.

**Comment by meroton-benjamin** [comment]
With the modifications after the pull request GetLogicalSize could be removed. Tests that verify functionality around the boundary conditions the  has been added (and the corresponding bug they discovered has been fixed).

**Comment by meroton-benjamin** [comment]
Sure I'll get on that

**Comment by meroton-benjamin** [comment]
Rebased quota enforced sector allocator on top of what is in main and made the tests pass. As the rebase was pretty big it might be worth making another review pass at this.

File: pkg/filesystem/pool/sector_map.go [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+
+type indirect[T any] struct {
+	val [indirectionArraySize]T
+}
+
+// GetNextMappedSector returns the logical index of the next mapped
+// sector in the sector map. If the start index is mapped then start is
+// returned. If there are no more mapped sectors then io.EOF is
+// returned.
+func (sp *SectorMap) GetNextMappedSector(logical uint64) (uint64, error) {
```
Comment (EdSchouten) [inline pkg/filesystem/pool/sector_map.go]: This method, but also many others, seem to return `error`. There are some methods which never fail, and in many cases we only return `nil` or `io.EOF`. What are your thoughts on getting rid of these `error` return values? Maybe in some cases we can use `bool`, while in others we can just return a really large/small/... number? In the case of `GetNextMappedSector()`, I see that the caller always calls `min()` against the result. Does that mean that just returning `tripleIndirectionSectors` would have had the same effect?
